File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int K = Integer.parseInt(st.nextToken());
        int N = Integer.parseInt(st.nextToken());

        int[] arr = new int[K]; // 각 랜선의 길이를 저장하기 위한 배열
        long max = 0;
        long min = 0;
        long mid = 0;

        for(int i = 0; i < K; i++) {
            int num = Integer.parseInt(br.readLine());
            arr[i] = num;

            if(max < num) { // 기존 최댓값이 더 작으면 재설정
                max = num;
            }
        }
        br.close();
        max++;

        while(min < max) {

            // 이분탐색 중간값 구하기
            mid = (max + min) / 2;

            long count = 0;

            for(int i = 0; i < arr.length; i++) { //
                count += (arr[i] / mid); // mid 길이로 자를 수 있는 갯수
            }

            if(count >= N) { // 목표 갯수보다 자른 수가 더 많으면 자르는 길이 증가(최솟값 증가)
                min = mid + 1;
            }
            else { // 자른 수가 더 적으면 자르는 길이 감소(최댓값 감소)
                max = mid;
            }
        }
        bw.write(min - 1 + "");
        bw.flush();
        bw.close();
    }
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.kt
import java.io.*
import java.util.*

var arr = arrayListOf<Int>()

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {

    val st = StringTokenizer(readLine())
    val k = st.nextToken().toInt() // 이미 가지고 있는 랜선의 갯수
    val n = st.nextToken().toInt() // 필요한 랜선의 갯수

    repeat(k) {
        arr.add(readLine().toInt())
    }

    arr.sort()

    close()

    println(binarySearch(n))
}

fun binarySearch(n : Int) : Long {
    var right : Long = arr[arr.lastIndex].toLong() + 1
    var left : Long  = 0
    var mid : Long

    while(left < right) {
        mid = (left + right) / 2
        var cnt : Long = 0

        for(elem in arr) {
            cnt += elem / mid
        }

        // 자른 랜선의 갯수가 부족하면 자를 길이를 줄여야 함
        if(cnt < n) {
            right = mid
        }
        // 자른 랜선의 갯수가 많으면 자를 길이를 늘려야 함
        else {
            left = mid + 1
        }
    }
    return left - 1
}